api/agent/drivers/docker/cookie.go

```diff
@@ -309,7 +309,12 @@ func (c *cookie) PullImage(ctx context.Context) error {
 		code := http.StatusInternalServerError
 		if dErr, ok := err.(*docker.Error); ok {
 			msg = dockerMsg(dErr)
-			code = dErr.Status // 401/404
+			code = dErr.Status // 500 or 404 according to docker-api
+
+			// This is unfortunate, docker-api returns "API error (500): unauthorized: authentication required" not 401.
+			if dErr.Status == http.StatusInternalServerError && strings.Contains(msg, "unauthorized: authentication required") {
```

Comment:
I think different docker registries (that is, different implementations of the docker api) might return different strings here. I'm not sure the api specification absolutely requires the message to be like that.

Reply:
it does not, and returns various codes for different modes of saying 'not authenticated' even, good catch